### PR TITLE
[SYCL][Joint Matrix] Implement code split for joint_matrix and joint_matrix_mad

### PR DESCRIPTION
**Pull request by dm-vodopyanov**
This patch implements device code split for `joint_matrix` and `joint_matrix_mad`.

Note: the data inside metadata values is sorted, so the order of matrix initialization in the C++ code is not relevant.

### CODE REVIEW

**Comment by dm-vodopyanov** [comment]
@intel/dpcpp-tools-reviewers friendly ping

File: llvm/test/tools/sycl-post-link/device-code-split/per-joint-matrix-3.ll [hunk 1]
```diff
@@ -0,0 +1,67 @@
+; This test is intended to check that we do not perform per-joint-matrix
+; split if it was disabled through one or another sycl-post-link option
+
+; RUN: sycl-post-link -symbols -S %s -o %t.table
+; RUN: FileCheck %s -input-file=%t.table --check-prefix CHECK-TABLE
+; RUN: FileCheck %s -input-file=%t_0.ll --check-prefix CHECK-IR
+;
+; -lower-esimd is needed so sycl-post-link does not complain about no actions
```
Comment (maksimsab) [inline llvm/test/tools/sycl-post-link/device-code-split/per-joint-matrix-3.ll:8]: We might want to have particular and explicit command line options doing enabling/disabling. For example, `-split-joint-matrix`, `-no-split-joint-matrix`. With these options we could perform this test in the following way:
`sycl-post-link -split=source -no-split-joint-matrix`.
It might look messy because we would have different options for splitting (`-split` vs `-split-joint-matrix`). However, using `-lower-esimd` is a dirty hack as well.

Please, share your thoughts.

**Review by maksimsab — CHANGES_REQUESTED** [review]

File: llvm/test/tools/sycl-post-link/device-code-split/per-joint-matrix-3.ll (hunk 1 above)
Comment (dm-vodopyanov) [reply llvm/test/tools/sycl-post-link/device-code-split/per-joint-matrix-3.ll:8]: Will we use these suggested options in production? If yes, in which cases? Or is this only for tests? If we will not use these options in production, I think it is better not to implement them

Comment (maksimsab) [reply llvm/test/tools/sycl-post-link/device-code-split/per-joint-matrix-3.ll:8]: Only for testing, right. We have a ton of things made only for testing. Why should it be an exception?

Comment (maksimsab) [reply llvm/test/tools/sycl-post-link/device-code-split/per-joint-matrix-3.ll:8]: For example, `-split-joint-matrix` might be the option turned on by default, and `-no-split-joint-matrix` might the option for testing.

Comment (AlexeySachkov) [reply llvm/test/tools/sycl-post-link/device-code-split/per-joint-matrix-3.ll:8]: We already have a few tests which use `-lower-esimd` (or any other action) to make `sycl-post-link` do something for a test.

I would be *against* adding a special flag for matrix for just testing with the following motivation:
- apparently we have a bit of disconnect with overall action-based design of the tool and its implementation where some features are not really implemented as separate high-level action. If we want to fix that, let's fix this as a separate refactoring PR, it is not unique for matrix feature
- the whole action-based approach seems weird to me from a production point of view: take device global as an example. Due to the way clang driver works, it cannot analyze the code in advance to see if there are any uses of device global to conditionally enable/disable their handling by `sycl-post-link`. Instead, it just always unconditionally passes the flag, wasting resources to allocate that string in memory and then pass it to an external process, which also parses it, etc.
- refactoring to `clang-linker-wrapper`. I imagine that at some point we will completely get rid of the tool and it will be a part of `clang-linker-wrapper` and most likely there will be a different approach for handling those which won't involve passing those "action" options around at all

To summarize: I don't think that it worth to solve this issue at the moment at all.

**Review by maksimsab — APPROVED** [review]
LGTM

**Comment by asudarsa** [comment]
I have a high-level question. Why do we need to split the code here? Can that be explained in the PR description please?

Thanks

**Comment by dm-vodopyanov** [comment]
> I have a high-level question. Why do we need to split the code here? Can that be explained in the PR description please?
> 
> Thanks

Added to the description.
